python/pydantic_core/core_schema.py

```diff
@@ -4013,3 +3811,84 @@ def definition_reference_schema(
 
 def _dict_not_none(**kwargs: Any) -> Any:
     return {k: v for k, v in kwargs.items() if v is not None}
+
+
+###############################################################################
+# All this stuff is deprecated by #980 and will be removed eventually
+# They're kept because some code external code will be using them
```

Comment:
I think we should just remove these and make it a breaking change. Bump the major version if you think we need to (I don't think we do, I'd like to keep considering `pydanic-core` as a quasi implementation detail)

Reply:
I think we should keep these since there will be some reasonable external use of them, and a deprecating warning is a lot better than an `ImportError`.

Reply:
but I'll put the attributes behind `__getattr__` so there's a warning.

Reply:
yes there is, the original issues that reported this used `FieldValidationInfo`, also:

* [here](https://github.com/jxnl/instructor/blob/49f23e9e5f1a4f7ccd79f75a1dbaa97761f2cee3/instructor/dsl/citation.py#L61)
* [here](https://github.com/DaloroAT/first_breaks_picking/blob/25095d13a2f27e4cd685b65ad0bde9e6d39e111d/first_breaks/data_models/dependent.py#L4)

and more, see [this github search](https://github.com/search?q=FieldValidationInfo+lang%3Apython&type=code)

Reply:
Hmm okay, you've convinced me there's some usage. I guess we can leave this complexity in here, such a bummer.

Reply:
I also used functions from the `core_schema.*` namespace [when adapting Pydantic V2 support for immoney](https://github.com/antonagestam/immoney/blob/e8fe2f00838e68cb487b199385540ae79b22b5bf/src/immoney/_pydantic.py#L337), and now quite surprised to see this usage is already deprecated (I haven't even had time to fully adopt V2 across libs yet, and API is already changing). What would have been the preferred option to use here, is there any way I could have avoided using these?

Reply:
We'll, we had to make this change to make pydantic V2 more compatible with v1 while still fully typed.

AFAIK this is the only post V2 change where we introduced incompatible changes.

It's unfortunate, but I don't see how we could have avoided it, or you could have predicted it.

Reply:
Gotcha, thanks for providing insight. Unfortunate and some extra churn for me, but I assume it's for the better in the long run 👍